homeassistant/components/huum/climate.py

```diff
@@ -0,0 +1,127 @@
+"""Support for Huum wifi-enabled sauna."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from huum.const import SaunaStatus
+from huum.exceptions import SafetyException
+from huum.huum import Huum
+
+from homeassistant.components.climate import (
+    ClimateEntity,
+    ClimateEntityFeature,
+    HVACMode,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_TEMPERATURE, PRECISION_WHOLE, UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Huum sauna with config flow."""
+    huum_handler = hass.data.setdefault(DOMAIN, {})[entry.entry_id]
+
+    async_add_entities([HuumDevice(huum_handler, entry.entry_id)], True)
+
+
+class HuumDevice(ClimateEntity):
+    """Representation of a heater."""
+
+    _attr_hvac_modes = [HVACMode.HEAT, HVACMode.OFF]
+    _attr_supported_features = ClimateEntityFeature.TARGET_TEMPERATURE
+    _attr_target_temperature_step = PRECISION_WHOLE
+    _attr_temperature_unit = UnitOfTemperature.CELSIUS
+    _attr_max_temp = 110
+    _attr_min_temp = 40
+    _attr_has_entity_name = True
+    _attr_name = None
+
+    _target_temperature = 40
+
+    def __init__(self, huum_handler: Huum, unique_id: str) -> None:
+        """Initialize the heater."""
+        self._attr_unique_id = unique_id
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, unique_id)},
+            name="Huum sauna",
+            manufacturer="Huum",
+        )
+
+        self._huum_handler = huum_handler
+        self._status = None
+
+    @property
+    def hvac_mode(self) -> HVACMode:
+        """Return hvac operation ie. heat, cool mode."""
+        if self._status and self._status.status == SaunaStatus.ONLINE_HEATING:
+            return HVACMode.HEAT
+        return HVACMode.OFF
+
+    @property
+    def icon(self) -> str:
+        """Return nice icon for heater."""
+        if self.hvac_mode == HVACMode.HEAT:
+            return "mdi:radiator"
+        return "mdi:radiator-off"
+
+    @property
+    def current_temperature(self) -> float | None:
+        """Return the current temperature."""
+        return getattr(self._status, "temperature", None) or self._attr_min_temp
```

Comment:
What is the third attribute of the `getattr`? If `self._status.temperature` does not exist, (without the third attribute) what does it return?

Also, otherwise we return the minimum temperature, but if the device can't report a current temperature, wouldn't it be better to return None as we don't know the current temperature?

Reply:
The third value is the default value, there is no point in ever using `getattr` without a third argument.

So if `self._status.temperature` is not set, the `getattr` call will return `None` which in turn will trigger the `or` and return `self._attr_min_temp`.

And we can return `None` as well yes, then I'll just remove the `or` statement

Reply:
But what i mean is,

```py
getattr(self._status, "temperature") == getattr(self._status, "temperature", None)
```

Is this true or false

Reply:
Just looked it up, that's not true, left would raise AttributeError.

Also wondering, if we can type `self._status` can we just return `self._status.temperature` directly?

Reply:
Exactly, it will raise `AttributeError`. Using `getattr` without a default value will is exactly the same as not using it at all.

So doing `self._status.temperature` is exactly the same as doing `getattr(self._status, "temperature")`, using `getattr` will do nothing here. There is actually a Ruff check for this as well, but I can't remember the plugin that needs to be enabled for it.

Reply:
Typing `_status` will not set any default values, it will just set the type of the variable, so I don't see how that would help.

The type of `_status` is `_status: HuumStatusResponse | None = None`, only after `async_update` has been called can we be sure that `_status` is not `None`, meaning that we have to type it as `None` as well since `async_update` can't be called before the constructor has been called.

Reply:
I mean that sounds good to me, just have a None check in the places where you return a value

Reply:
Refactored

Reply:
Here is the `ruff` rule that I mentioned earlier btw 🙂  Might be worth enabling for the entire codebase at some point.